redis/keyspace_utils.cc

```diff
@@ -0,0 +1,210 @@
+     "save SETs for redis"
+    )));
+    builder.set_gc_grace_seconds(0);
+    builder.with(schema_builder::compact_storage::yes);
+    builder.with_version(db::system_keyspace::generate_schema_version(builder.uuid()));
+    return builder.build(schema_builder::compact_storage::yes);
+}
+
+schema_ptr zsets_schema(sstring ks_name) {
+     schema_builder builder(make_lw_shared(schema(generate_legacy_id(ks_name, redis::ZSETs), ks_name, redis::ZSETs,
+     // partition key
+     {{"pkey", utf8_type}},
+     // clustering key
+     {{"ckey", double_type}},
+     // regular columns
+     {{"data", utf8_type}},
+     // static columns
+     {},
+     // regular column name type
+     utf8_type,
+     // comment
+     "save ZSETs for redis"
+    )));
+    builder.set_gc_grace_seconds(0);
+    builder.with(schema_builder::compact_storage::yes);
+    builder.with_version(db::system_keyspace::generate_schema_version(builder.uuid()));
+    return builder.build(schema_builder::compact_storage::yes);
+}
+
+future<> create_keyspace_if_not_exists_impl(lw_shared_ptr<db::config> config, int default_replication_factor) {
+    auto keyspace_options = config->redis_keyspace_options();
+    if (keyspace_options.count("class") == 0) {
+        keyspace_options["class"] = "SimpleStrategy";
+        keyspace_options["replication_factor"] = sprint("%d", default_replication_factor);
+    }
+    auto keyspace_gen = [config, keyspace_options = std::move(keyspace_options)]  (sstring name) {
+        auto& proxy = service::get_local_storage_proxy();
+        if (proxy.get_db().local().has_keyspace(name)) {
+            return make_ready_future<>();
+        }
+        auto attrs = make_shared<cql3::statements::ks_prop_defs>();
+        attrs->add_property(cql3::statements::ks_prop_defs::KW_DURABLE_WRITES, "true");
+        std::map<sstring, sstring> replication_properties;
+        for (auto&& e : keyspace_options) {
+            replication_properties.emplace(e.first, e.second);
+        }
+        attrs->add_property(cql3::statements::ks_prop_defs::KW_REPLICATION, replication_properties); 
+        attrs->validate();
+        const auto& tm = service::get_local_storage_service().get_token_metadata();
+        return service::get_local_migration_manager().announce_new_keyspace(attrs->as_ks_metadata(name, tm), false);
+    };
+    auto table_gen = [] (sstring ks_name, sstring cf_name, schema_ptr schema) {
+        auto& proxy = service::get_local_storage_proxy();
+        if (proxy.get_db().local().has_schema(ks_name, cf_name)) {
+            return make_ready_future<>();
+        }
+        logger.info("Create keyspace: {}, table: {} for redis.", ks_name, cf_name);
+        return service::get_local_migration_manager().announce_new_column_family(schema, false);
+    };
+    // create 16 default database for redis.
+    return parallel_for_each(boost::irange<unsigned>(0, 16), [keyspace_gen = std::move(keyspace_gen), table_gen = std::move(table_gen)] (auto c) {
```

Comment:
SELECT command is supported, it would switch between keyspaces. The limitation is got from configuration.

So if we config redis_default_database_count to be larger than 16, SELECT allows to switch to unexisting keyspace, and it will fail.